libgalois/include/katana/PropertyGraph.h

```diff
@@ -767,6 +819,620 @@ CreateSymmetricGraph(PropertyGraph* pg);
+
+  EdgeLabelTy label(uint32_t index) const noexcept {
+    KATANA_LOG_ASSERT(index < num_unique_edge_labels_);
+    KATANA_LOG_ASSERT(size_t(index) < edge_index_to_label_map_.size());
+    return edge_index_to_label_map_[index];
+  }
+
+  uint32_t index(const EdgeLabelTy& label) const noexcept {
+    KATANA_LOG_ASSERT(edge_label_to_index_map_.count(label) > 0);
+    // return edge_label_to_index_map_[label];
+    return edge_label_to_index_map_.find(label)->second;
+  }
+
+  uint32_t num_unique_labels() const noexcept {
+    return num_unique_edge_labels_;
+  }
+
+  /**
+   * @param label: label to check
+   * @returns true iff there exists some edge in the graph with that label
+   */
+  bool has_edge_label(const EdgeLabelTy& label) const noexcept {
+    return (
+        edge_label_to_index_map_.find(label) !=
+        edge_label_to_index_map_.cend());
+  }
+#if 0
+  /**
+   * Wrapper to get the begin iterator to distinct edge labels in the graph.
+   *
+   * @returns Iterator to first distinct edge label
+   */
+  auto DistinctEdgeLabelsBegin() const noexcept {
+    return edge_index_to_label_map_.cbegin();
+  }
+
+  /**
+   * Wrapper to get the end iterator to distinct edge labels in the graph.
+   *
+   * @returns Iterator to end distinct edge label
+   */
+  auto DistinctEdgeLabelsEnd() const noexcept {
+    return edge_index_to_label_map_.cend();
+  }
+#endif
+
+  /**
+   * Wrapper to get the distinct edge labels in the graph.
+   *
+   * @returns Range of the distinct edge labels
+   */
+  auto distinct_edge_labels() const noexcept {
+    return MakeStandardRange(
+        edge_index_to_label_map_.cbegin(), edge_index_to_label_map_.cend());
+  }
+
+private:
+  // map an integer id to each unique edge label in the graph
+  using EdgeLabelToIndexMap = std::unordered_map<EdgeLabelTy, uint32_t>;
+  // each index refers to a unique edge label in the graph
+  using EdgeIndexToLabelMap = std::vector<EdgeLabelTy>;
+
+  EdgeLabelIndex(
+      EdgeLabelToIndexMap&& edge_label_to_index,
+      EdgeIndexToLabelMap&& edge_index_to_label, uint32_t num_labels) noexcept
+      : edge_label_to_index_map_(std::move(edge_label_to_index)),
+        edge_index_to_label_map_(std::move(edge_index_to_label)),
+        num_unique_edge_labels_(num_labels) {
+    KATANA_LOG_DEBUG_ASSERT(
+        num_labels == edge_index_to_label_map_.size() &&
+        num_labels == edge_label_to_index_map_.size());
+  }
+
+  EdgeLabelToIndexMap edge_label_to_index_map_;
+  EdgeIndexToLabelMap edge_index_to_label_map_;
+  uint32_t num_unique_edge_labels_ = 0u;
+};
+
+class KATANA_EXPORT EdgeLabelAwareTopology;
+
+// TODO(amber): make OrigEdgeIDMap optional via template argument
+class KATANA_EXPORT TransposeTopology : public GraphTopology {
+  using OrigEdgeIDMap = katana::NUMAArray<Edge>;
+
+public:
+  TransposeTopology() = default;
+  TransposeTopology(TransposeTopology&&) = default;
+  TransposeTopology& operator=(TransposeTopology&&) = default;
```

Comment:
Don't move semantics not play nice with subclassing? I feel like we might need to use pointers and references to allow functions to take an arbitrary `GraphTopology` without a template parameter. Which is important for exposing this to Python. Sorry to repeat myself.

Reply:
I don't see why move semantics won't work with simple inheritance.

Reply:
Because you cannot move a value of type `TransposeTopology` into a variable of type `GraphTopology`. This would prevent code that was not templatized from performing moves abstractly. There may be a way around this, so that non-templatized code never needs to perform a move, but it seems like it could cause problems.